### ISSUE TITLE
[FEA] Read JSON support for dask_cudf

### ISSUE THREAD

**Issue by taureandyernv** [rapidsai/dask-cudf#253]
We're seeing a repeated request form the commuity to have read JSON support, esp for distributed systems.  Can we implement it please?

**Comment by quasiben** [comment]
Sure, I think `cudf.read_json` was merged earlier today: https://github.com/rapidsai/cudf/pull/1630

**Comment by mrocklin** [comment]
@rjzamora would you be interested in this?  This would get into into cudf a bit more, and would also force you to learn about how Dask handles computations internally.  

These PRs are good examples to work from:

-  https://github.com/rapidsai/dask-cudf/pull/58/files
-  https://github.com/rapidsai/dask-cudf/pull/220/files

Relevant docs: 

-  https://docs.dask.org/en/latest/custom-graphs.html
-  https://docs.dask.org/en/latest/dataframe-design.html

**Comment by mrocklin** [comment]
Actually, as a first pass, we might consider using Dask's internal bytes functionality.  This is probably less efficient, because we'll be grabbing data using normal methods and going through host memory, but it's a bit more robust, handles cloud object stores, and so on.  It might be totally fine performance-wise.

So rather than the two PRs listed above we might try seeing if we can reuse this code:

https://github.com/dask/dask/blob/master/dask/dataframe/io/json.py#L77-L167

And in particular maybe we can just make the `pandas.read_json` function a keyword argument to that function, and then pass in `cudf.read_json`.  Then we might be able to reuse the existing function.

```python
import dask.dataframe as dd
df = dd.read_json('/path/to/files.*.json', read_json=cudf.read_json)
```

This might *just work* today if the cudf.read_json function handles the things that are expected of it here: 

```python
    df = pd.read_json(s, orient='records', lines=True, **kwargs)
```

**Comment by mrocklin** [comment]
So this would require a patch to `dask`, and then probably a very small patch to dask-cudf, which just adds a one-line function that defaults `cudf.read_json`.

**Comment by rjzamora** [comment]
Sounds good @mrocklin - I'll start looking at this today

**Comment by rjzamora** [comment]
@mrocklin I am thinking through the changes you suggested, and the first recommendation makes sense to me (I think):

1. change the [`read_json_chunk`](https://github.com/dask/dask/blob/edec11047c999a40c387c657c6811ab00c655a89/dask/dataframe/io/json.py#L170)  and `read_json_file` definitions to do something like this:

```
def read_json_file(f, orient, lines, kwargs):
    read_function = kwargs.pop('read_function', pd.read_json)
    with f as f:
        return read_function(f, orient=orient, lines=lines, **kwargs)
```
2. Use:  `df = dd.read_json(f, read_function=cudf.read_json)` in user app

The suggested change in  `dask-cudf` is a little fuzier to me. Would you like to add a `json.py` file to `dask-cudf` with a `read_json` function that uses `cudf.read_json`?

**Comment by mrocklin** [comment]
1.  Seems like we're on the same page

2.  Yes, for user convenience, but we can visit that after we've verified that the approach in 1 works well

**Comment by rjzamora** [comment]
Ok, (1) works fine with a small change in `cudf` (`cudf.read_json` expects a path, but dask wants to pass along a file handle).  My thought is to do this in `cudf`:

```p
if hasattr(path_or_buf, 'name'):
    path_or_buf = path_or_buf.name
df = cpp_read_json(path_or_buf, dtype, lines, compression, byte_range)
```

**Comment by mrocklin** [comment]
I think that `cudf.read_csv` was able to take a Python `io.BytesIO` object.  Perhaps `cudf.read_json` does as well?  If not, then maybe that's a change that we could ask for.

**Comment by kkraus14** [comment]
I think the first step should be to open an issue for discussion / tracking in cuDF regardless. If it doesn't accept the standard input objects then it should be updated to do so. 😄

**Comment by mrocklin** [comment]
Yeah, looking at the code it looks like the name `path_of_buf` was intended to imply this ability.  It's definitely there in read_csv, but also definitely not in read_json.  My *guess* is that it's an easy-ish fix for the cuio folks.  +1 on raising on cudf and asking for a bit of help.  My guess is that they'll have things done in a short time, but they'll have a better sense of things.

**Comment by rjzamora** [comment]
Right, the csv bindings in cudf use an `is_file_like` function to handle different input types. Raising an issue to suggest the same approach for json makes sense to me - I can do this if someone hasn't already

**Comment by rjzamora** [comment]
I raised a new issue in cudf about this (https://github.com/rapidsai/cudf/issues/1769) - In the meantime, I will submit a PR to add a `read_json` function in `dask_cudf`.  For now, this function can just copy most of the dask code (but will use `cudf.read_json` when appropriate).

**Comment by mrocklin** [comment]
> I will submit a PR to add a read_json function in dask_cudf. For now, this function can just copy most of the dask code (but will use cudf.read_json when appropriate).

Ah, let's avoid copying the code if we can.  I can see how this could be construed from my point above though.

For point 2 above what I meant was that we might have a function like

```python
read_json = functools.partial(dask.dataframe.read_json, read_function=cudf.read_json)
```

Or, perhaps less fancily

```python
def read_json(...):
    return dd.read_json(..., read_function=cudf.read_json)
```

I wouldn't expect any more code in dask-cudf than that.

**Comment by rjzamora** [comment]
@mrocklin I agree with this approach once dask is changed to handle the `read_function` keyword`, and cudf is modified to handle non-string input parameters for the file path.

The "quick fix" I suggested is just a temporary solution that does not require changes in dask or cudf -  If you think we should wait for action in dask and cudf before addding anything in dask-cudf, I certainly understand.  Let me know if I should pause my efforts on this PR: https://github.com/rapidsai/dask-cudf/pull/255

**Comment by mrocklin** [comment]
I think that we should wait.  I think that changes in both upstream
repositories will take no more than a few days.  We have plenty of other
work to keep us occupied in the meantime I think.

**Comment by rjzamora** [comment]
That works for me - will update the existing PR with the streamlined solution once it makes sense

**Comment by quasiben** [comment]
https://github.com/rapidsai/dask-cudf/pull/255 should resolve this PR

**Comment by randerzander** [comment]
Closed by #255 , @taureandyernv let us know if this isn't working for your usecases, please!
